bootstore/src/schemes/v0/fsm.rs

```diff
@@ -0,0 +1,479 @@
+    ) -> Output {
+        let State::Uninitialized(_) = self.state.as_ref().unwrap() else {
+            return ApiError::RackAlreadyInitialized.into();
+        };
+        let total_members = initial_membership.len();
+        match create_pkgs(rack_uuid, initial_membership.clone()) {
+            Ok(pkgs) => {
+                // We learn our pkg by iterating
+                let mut our_pkg = None;
+                let request_id = Uuid::new_v4();
+                let envelopes: Vec<_> = pkgs
+                    .expose_secret()
+                    .into_iter()
+                    .zip(initial_membership)
+                    .filter_map(|(pkg, peer)| {
+                        if peer == self.common.id {
+                            // Don't send a message to ourself
+                            our_pkg = Some(pkg.clone());
+                            None
+                        } else {
+                            Some(Envelope {
+                                to: peer,
+                                msg: Request {
+                                    id: request_id,
+                                    type_: RequestType::Init(pkg.clone()),
+                                }
+                                .into(),
+                            })
+                        }
+                    })
+                    .collect();
+                self.state = Some(State::InitialMember(InitialMemberState {
+                    pkg: our_pkg.unwrap(),
+                    distributed_shares: BTreeMap::new(),
+                    rack_init_state: Some(RackInitState::Running {
+                        start: self.common.clock,
+                        total_members,
+                        acks: BTreeSet::from([self.common.id.clone()]),
+                        unacked_envelopes: envelopes.clone(),
+                    }),
+                    pending_learn_requests: BTreeMap::new(),
+                }));
+                // Note that while we could only send the envelopes to
+                // connected peers, the network layer will just drop them.
+                // We already have a copy to send on connection stored in
+                // `Rack::InitState::Running.unacked_envelopes`.
+                Output { persist: true, envelopes, api_output: None }
+            }
+            Err(e) => return ApiError::RackInitFailed(e).into(),
+        }
+    }
+
+    /// Initialize a node added after rack initialization
+    pub fn init_learner(&mut self) -> Output {
+        let State::Uninitialized(_) = self.state.as_ref().unwrap() else {
+            return ApiError::PeerAlreadyInitialized.into();
+        };
+
+        let mut state = LearningState { attempt: None };
+        let mut output = state.new_attempt(&mut self.common);
+        self.state = Some(State::Learning(state).into());
+        // We changed state into a learner, so persist this knowledge
+        output.persist = true;
```

Comment:
It looks like this function is the only caller of `new_attempt`; should `new_attempt` just set the `persist` variable to true itself?

Reply:
No longer relevant as of [98ada27](https://github.com/oxidecomputer/omicron/pull/3385/commits/98ada27b53d4429463cde94cc91bedec0bc4ef98)